[JENKINS-33663] - Upgrade wizard

kzantow opened the pull request:
Adapt the setup wizard GUI to provide a similar experience upgrading Jenkins.

This modifies / enhances the original banner to drop into an upgrade wizard with information about what's new in Jenkins 2 and providing a way to get the base set of plugins updated.

This addresses: https://issues.jenkins-ci.org/browse/JENKINS-33663

! NOTE: many of these files have existing whitespace issues, I plan on cleaning them all up after having the changes reviewed here

Screenshots:

![image](https://cloud.githubusercontent.com/assets/3009477/14825079/74ff045c-0ba6-11e6-8b49-1e49458d481c.png)

![image](https://cloud.githubusercontent.com/assets/3009477/15332417/794d2d48-1c33-11e6-85c6-b1c976fdb8ab.png)

![image](https://cloud.githubusercontent.com/assets/3009477/15404829/5c316f30-1dcd-11e6-9268-1ddad9cd31b3.png)

![image](https://cloud.githubusercontent.com/assets/3009477/15332534/fe54f336-1c33-11e6-9607-5818b11692e6.png)

amuniz commented:
Manually tried. The upgrade window appears but it's blank. Browser console errors:

```
http://localhost:8080/adjuncts/d77aebba/jenkins/install/css/icons/icomoon.css Failed to load resource: the server responded with a status of 404 (Not Found)
http://localhost:8080/setupWizard/loadTemplate?templateName=upgradeTo20Panel&_=1461315124799 Failed to load resource: the server responded with a status of 404 (Not Found)
http://localhost:8080/adjuncts/d77aebba/jenkins/install/css/icons/icomoon.css Failed to load resource: the server responded with a status of 404 (Not Found)
```

I guess you already knew this (and that's why it's a WiP?).

amuniz commented on core/src/main/java/jenkins/install/InstallState.java:
```diff
@@ -25,63 +25,88 @@
      * Need InstallState != NEW for tests by default
      */
-    UNKNOWN(true, null),
+    @Extension
```
`@Extension` in a class field? What's the effect of this?

amuniz replied:
I guess you want all this objects into the extensions container?

kzantow replied:
Yes, exactly. I want the instances. It seems to work (and is documented to work this way) @amuniz

amuniz commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -4587,6 +4580,26 @@ private static void computeVersion(ServletContext context) {
         }
         String ver = props.getProperty("version");
         if(ver==null)   ver = UNCOMPUTED_VERSION;
+        if(Main.isDevelopmentMode && "${build.version}".equals(ver)) {
```
Why the second condition? Is `isDevelopmentMode` not enough?

kzantow replied:
usually this is set, some cases (ahem debugging using eclipse) it is not

amuniz commented on war/src/main/js/pluginSetupWizardGui.js:
```diff
@@ -333,14 +336,38 @@ var createPluginSetupWizard = function(appendTarget) {
+	
+	// indicates 
+	var transitionPanel = function(state) {
+		if(/CREATE_ADMIN_USER/.test(state)) {
```
Is `===` not valid here?

kzantow commented on core/src/main/java/jenkins/install/InstallState.java:
```diff
@@ -25,63 +25,88 @@
  *
  * @author <a href="mailto:tom.fennelly@gmail.com">tom.fennelly@gmail.com</a>
  */
-@Restricted(NoExternalUse.class)
```
This may be referenced by extensions to add more install states

kzantow commented on core/src/main/java/jenkins/install/InstallUtil.java:
```diff
@@ -74,7 +74,7 @@
     public static InstallState getInstallState() {
         // Support a simple state override. Useful for testing.
-        String stateOverride = System.getenv("jenkins.install.state");
+        String stateOverride = System.getProperty("jenkins.install.state", System.getenv("jenkins.install.state"));
```
Easier to do -Djenkins.install.state=upgrade if doing development

svanoort replied:
I like this.  A lot.

jtnord replied:
Windows development!  because this will not work on [Linux with Bash](http://unix.stackexchange.com/questions/93532/exporting-a-variable-with-dot-in-it#answer-93533) as you can not have environment variables with dots in them :-)

kzantow commented on core/src/main/java/jenkins/install/SetupWizard.java:
```diff
@@ -55,79 +65,81 @@
-                iapf.touch(System.currentTimeMillis());
-                iapf.chmod(0640);
-                iapf.write(randomUUID + System.lineSeparator(), "UTF-8");
+        if(newInstall) {
```
this is the only real change here, moved some of these things to only happen for a new install, rest is whitespace

amuniz commented:
Re-tested after last modifications. All my manual tests passed.

Just one question; I saw only "Install Recommended Plugins" option available, is that right? Why I could not select what to install?

And one detail (but not sure if related to this PR); there is an empty space over the breadcrumb (is that intentional):

<img width="813" alt="screenshot1" src="https://cloud.githubusercontent.com/assets/1017585/14801312/a27bba32-0b48-11e6-9f50-b1ee3d29b64c.png">

kzantow commented:
@amuniz the only options are to install the recommended plugins or decline. presumably they already have plugins installed and the plugin list isn't yet built to handle already installed plugins in a meaningful manner. the idea here is to get the recommended plugins installed as a baseline across all the Jenkins installations, or the existing ones will be missing out on some of the highlights of Jenkins 2. Regarding the space on the page, it's fixed, it was definitely an issue caused by this PR.

amuniz commented:
Ok, thanks.
This is :+1: for me

svanoort commented on core/src/main/java/jenkins/install/SetupWizard.java:
```diff
@@ -208,14 +263,30 @@ public HttpResponse doCompleteInstall() throws IOException, ServletException {
+            jenkins.setInstallState(InstallState.CONFIGURE_SECURITY.getNextState());
+            return HttpResponses.okJSON();
+        }
+        return HttpResponses.errorJSON("Unable to configure security");
```
Possibly a naive question here, but is there anything we need to revert in the event of a failure to configure the security?

kzantow replied:
@svanoort not really sure there's anything we can do in this case. this code wasn't changed, however just indented

svanoort replied:
@kzantow Yeah, I'm not exactly sure where this was originally touched, I'm just seeing it for the first time now and it's making my Spidey Sense tingle.  Was this code path present before we added the SetupWizard somehow?

kzantow replied:
@svanoort oh, wait a sec. yes, this was added as one of the derivations of the setup wizard, where security configuration is included as one of the steps... but then removed, now temporarily re-added as part 2 of the upgrade wizard is to deal with security, sorry I was talking about the other security saving spot re: indentation.

kzantow replied:
... falling victim to the 'staring at this code too much' syndrome

svanoort commented on core/src/main/java/jenkins/install/UpgradeWizard.java:
```diff
@@ -101,6 +109,38 @@ public boolean isDue() {
+    public boolean isShowUpgradeWizard() {
+        HttpSession session = Stapler.getCurrentRequest().getSession(false);
+        if(session != null) {
+            return Boolean.TRUE.equals(session.getAttribute(SHOW_UPGRADE_WIZARD_FLAG));
```
I'm confused by this boolean test: won't this ALWAYS return false when comparing against a string?

svanoort commented on war/src/main/js/util/jenkins.js:
```diff
@@ -184,12 +184,36 @@ exports.loadTranslations = function(bundleName, handler, onError) {
 /**
+ * Load a handlebars template from the server
+ */
+exports.loadTemplate = function(templateName, handler, onError) {
```
Maybe an 🐜  but I have to wonder if there isn't a better way to do this... perhaps by doing a little maven configuration to throw our JS/templates/etc resources in WEB-INF so they can be directly exposed from the WAR as simple URLs?  Feels like we're reinventing the wheel by writing a custom function to serve static content using an URL param vs. using parts of the Servlet spec.

kzantow replied:
Yeah, agree.. this is actually not being used anyway, it was a byproduct of my first idea to make this happen... I'll remove it.

svanoort replied:
@kzantow Excellent - the removal commit for this looks good, btw

rtyler commented on core/src/main/resources/jenkins/install/pluginSetupWizard.properties:
```diff
@@ -56,3 +58,17 @@ installWizard_pluginInstallFailure_message=Some plugins failed to install proper
+installWizard_upgradePanel_banner=Welcome to Jenkins 2!
+installWizard_upgradePanel_message=Jenkins 2 includes some great new features we think you'll love to use. Get your instance updated to the latest \
+and greatest plugins available and take advantage of them!
+installWizard_upgradePanel_recommendedActionTitle=Install Recommended Plugins
```
s/Recommended/Suggested/

The terminology should be consistent (suggested) between the upgrade and new user flow.

rtyler commented on core/src/main/resources/jenkins/install/pluginSetupWizard.properties:
```diff
@@ -56,3 +58,17 @@ installWizard_pluginInstallFailure_message=Some plugins failed to install proper
 installWizard_retry=Retry
+installWizard_upgradePanel_title=Upgrade
+installWizard_upgradePanel_banner=Welcome to Jenkins 2!
+installWizard_upgradePanel_message=Jenkins 2 includes some great new features we think you'll love to use. Get your instance updated to the latest \
```
"Get your instance updated" doesn't parse well IMHO. I suggest the text for this message to be:

> Jenkins 2 includes some great new features that we think you'll love, install these additional plugins to take advantage of them!

batmat replied:
@rtyler More a question from a non native speaker: didn't we say using the full form _you will_ instead of _you'll_ would be preferable in the UI texts? 

Possibly I saw some fixes along those lines in the recent PRs @orrc did about improving many helps and so on.

rtyler commented on core/src/main/resources/jenkins/install/pluginSetupWizard.properties:
```diff
@@ -56,3 +58,17 @@ installWizard_pluginInstallFailure_message=Some plugins failed to install proper
+installWizard_upgradePanel_skipRecommendedPlugins=No thanks
+installWizard_upgradeComplete_title=Upgrade
+installWizard_pluginsInstalled_banner=Welcome to Jenkins 2!
+installWizard_upgradeComplete_message=Congratulations! You've upgraded to Jenkins 2.</p><p>To learn more about the great new features, visit <a target="_blank" href="https://jenkins.io/2.0/">the Jenkins 2 website</a>!
```
> You've upgraded

Better just to write "You have upgraded", some people get weird about the contraction "you've"

batmat replied:
Hah, should have read that before maybe :)

rtyler commented on core/src/main/resources/jenkins/install/pluginSetupWizard.properties:
```diff
@@ -56,3 +58,17 @@ installWizard_pluginInstallFailure_message=Some plugins failed to install proper
+installWizard_upgradeComplete_message=Congratulations! You've upgraded to Jenkins 2.</p><p>To learn more about the great new features, visit <a target="_blank" href="https://jenkins.io/2.0/">the Jenkins 2 website</a>!
+installWizard_upgradeSkipped_title=Upgrade
+installWizard_upgradeSkipped_banner=Features Not Installed
+installWizard_upgradeSkipped_message=<div class="alert alert-warning fade in">You've decided to skip updating to the great features in Jenkins 2.</div></p><p><a target="_blank" href="https://jenkins.io/2.0/">Click here to learn what you're missing</a>!
```
> You've decided to skip updating to the great features in Jenkins 2. 


I suggest just dropping this sentence, it comes across almost patronizing to me. Just the click prompt should be sufficient IMO

jtnord commented on core/src/main/java/jenkins/install/SetupWizard.java:
```diff
@@ -216,6 +272,21 @@ static void completeSetup(Jenkins jenkins) {
+     * Handle security configuration
+     */
+    public HttpResponse doSaveSecurityConfig(StaplerRequest req, StaplerResponse rsp) throws IOException, ServletException, FormException {
+        jenkins.checkPermission(Jenkins.ADMINISTER); // this is redundant, maybe
```
this is not redundant....

daniel-beck commented:
* Once I click the button, the overlay appears on every URL and I have no way to postpone the choice. I can only install, or decline, but not postpone in any way if I understand this approach correctly?
* Showing the version number in the footer when upgrading makes no sense.
* I don't know, the wording is still pretty… passive-aggressive, I guess? "Click here to learn what you're missing", why not start with this information to begin with, so I can make an informed choice?

daniel-beck commented:
* The "X" close button on the banner (which postpones for a day) is different from the "No thanks" option in the wizard itself.
* Reusing the list of plugins from the initial setup makes no sense, as many of them are there to provide some basic functionality for Jenkins. If you're upgrading, it's safe to assume you already have that, so the 'exciting new features' should be limited to Pipeline and related plugins (and only if not already installed, surprise upgrades are not fun!), and not include Subversion plugin and LDAP plugin. There's also the problem of not being able to dynamically deploy because dependencies are already installed but in outdated versions.

That last item is a serious problem, so :-1: We haven't upgraded bundled Subversion Plugin beyond 2.0 for years because of the possible upgrade compatibility problem, and now we're doing that through the backdoor.

* A proper solution to the upgrade problem will also need to take care of [JENKINS-34494](https://issues.jenkins-ci.org/browse/JENKINS-34494), for which we already received a number of reports.

daniel-beck commented:
Some of the strings make no sense for an upgrade situation: "Start using Jenkins"?

oleg-nenashev commented on core/src/main/java/jenkins/install/InstallState.java:
```diff
@@ -24,64 +24,75 @@
  */
-@Restricted(NoExternalUse.class)
-public enum InstallState {
+public class InstallState implements ExtensionPoint {
```
2.0 has been released.
Now it becomes a binary compatibility issue.

@daniel-beck Do we have a policy regarding it till 2.0 LTS?

daniel-beck replied:
@oleg-nenashev It was `@Restricted`, so I don't understand what you're saying. This could as well be a different type, with larger changes to core to refer to the new type, and we could delete the old one.

oleg-nenashev replied:
Yeah, no problem with the Restricted stuff

daniel-beck commented:
@kzantow This needs merge conflicts resolved.

oleg-nenashev commented:
GitHub does not visualize it correctly, but in https://github.com/jenkinsci/jenkins/pull/2281/commits/6d657c4e42485a966ad55c1de844fa10d29f4092 there was a mistake during the merge operation. The bits from https://github.com/jenkinsci/jenkins/pull/2328 have not been integrated correctly. 

From what I see in the manual tests, https://github.com/jenkinsci/jenkins/blob/master/war/src/main/js/util/jenkins.js#L212-213 will be removed if we merge this stuff as is. In such case the update center will be showing the offline pane if an update site skips a connectivity check.

Hence 🐛 . I continue local tests

oleg-nenashev commented:
Removed bug, because it rather a jsbundles issue
"src//main/js/jsbundles/pluginWizard.js" is not being updated properly even during "mvn clean build". Smells like a bug in js-bundles (cc @tfennelly ). Had to delete the folder and rebuild it with Maven.

kzantow commented:
@oleg-nenashev #2328 is are merged in this, look the same as in the PR looking at my local copy

oleg-nenashev commented:
@kzantow 
Yes, we have discovered the issue together with @tfennelly . In my case it was a leftover from the nerge of https://github.com/jenkinsci/jenkins/pull/2322 into my local branch.

Now everything works fine. Manually tested the pluggability of plugin suggestions, it works like a charm.

oleg-nenashev commented:
I see that installation of particular selected plugins hangs the InstallationWizard. "installStatus" request returns the "CREATE_ADMIN_USER" state, so most likely the UI switched to the next step.

In the log I'm getting "pluginSetupWizard" already installed. Maybe another merge error

oleg-nenashev commented on core/src/main/java/jenkins/install/SetupWizard.java:
```diff
@@ -245,4 +270,170 @@ public String getRequestURI() {
+                
+                try {
+                    if(connection instanceof HttpURLConnection) {
+                        int responseCode = ((HttpURLConnection)connection).getResponseCode();
```
Nice2have: explicit timeout

kzantow replied:
@oleg-nenashev the proxy now defines a shorter global timeout, so things like this won't take forever

kzantow replied:
... for anything using `ProxyConfiguration.open()`, I should say

oleg-nenashev commented on core/src/main/java/jenkins/install/SetupWizard.java:
```diff
@@ -245,4 +270,170 @@ public String getRequestURI() {
+
+    
+    /**
+     * Gets the suggested plugin list from the update sites, falling back to a local version
```
NIT: <code>@since TODO</code>

oleg-nenashev commented on core/src/main/java/jenkins/install/SetupWizard.java:
```diff
@@ -45,89 +67,94 @@
+                    // most native packages of Jenkins creates a machine user account 'jenkins' to run Jenkins,
+                    // and use group 'jenkins' for admins. So we allow groups to read this file
+                    iapf.touch(System.currentTimeMillis());
+                    iapf.chmod(0640);
```
Noop on windows. Is it safe?

kzantow replied:
yes, this was debated way too much in a different pr, it's just been moved

oleg-nenashev commented on core/src/main/java/jenkins/install/SetupWizard.java:
```diff
@@ -245,4 +270,170 @@ public String getRequestURI() {
+        VersionNumber from = new VersionNumber("1.0");
+        File state = getUpdateStateFile();
+        if (state.exists()) {
+            from = new VersionNumber(defaultIfBlank(readFileToString(state), "1.0").trim());
```
I'm not sure about the behavior for blank files. Maybe IOException is preferable

kzantow replied:
There's another spot that does a `touch` on the file, even with no version written, this is how it's "snoozed" for 1 day

oleg-nenashev replied:
OK. a code comment or Javadoc could help

oleg-nenashev commented on core/src/main/java/jenkins/install/SetupWizard.java:
```diff
@@ -245,4 +270,170 @@ public String getRequestURI() {
+        return Jenkins.getInstance().getInstallState().getStartPanel(new Provider<String>() {
+            @Override
+            public String get() {
+                return "welcomePanel";
```
It could help the extensibility cases if we can change this value via system property.

kzantow replied:
... but we can just override this in the install state, and override which install state is active

oleg-nenashev replied:
agreed

oleg-nenashev commented:
Manual testing of the PR was successful. 
I see no obvious issues in the code, hence 🐝

oleg-nenashev commented:
@daniel-beck Do you still keep a bug?

mslusarczyk commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -1234,8 +1236,11 @@ private void trackInitialPluginInstall(@Nonnull final List<Future<UpdateCenter.U
-            jenkins.setInstallState(InstallState.INITIAL_PLUGINS_INSTALLING);
+        if (!Jenkins.getInstance().getInstallState().isSetupComplete()) {
+            if (jenkins.getInstallState() == InstallState.NEW) {
+                jenkins.setInstallState(InstallState.INITIAL_PLUGINS_INSTALLING);
```
why this is not a part of InstallState proceed-chain? Seems that transition from NEW to INITIAL_PLUGINS_INSTALLING is done in some custom way.

kzantow replied:
yeah, this was done before I changed to use `InstallState.proceed...` I'll make sure this gets updated

oleg-nenashev commented:
Seems it's WiP

mslusarczyk commented on core/src/main/java/jenkins/install/InstallUtil.java:
```diff
@@ -70,12 +75,60 @@
+    /**
+     * Returns the next state during a transition from the current install state
+     */
+    public static InstallState getNextInstallState() {
```
unused method.

mslusarczyk commented on core/src/main/resources/jenkins/install/UpgradeWizard/client-scripts.jelly:
```diff
@@ -53,12 +52,12 @@
     <div style="display:none">
       <div id="upgrade-fragment">
-        <a id="upgrade-snooze" class="post" href="${rootURL}/${it.url}/snooze">
+        <a id="upgrade-snooze" class="post" href="${rootURL}/installState/snooze">
```
get we get 'installState' from some kind of const?

mslusarczyk commented on war/src/main/js/templates/firstUserPanel.hbs:
```diff
@@ -7,9 +7,6 @@
 	</div>
 </div>
 <div class="modal-footer">
-    <button type="button" class="btn btn-link skip-first-user" disabled>
```
'skip' button moved or removed?

mslusarczyk commented on war/src/main/js/util/jenkins.js:
```diff
@@ -184,8 +184,22 @@ exports.loadTranslations = function(bundleName, handler, onError) {
+					if(property in target) {
+						return target[property];
+					}
+					console.log('"' + property + '" not found in translation bundle.');
```
console.log in prod code? maybe if(debug) would be a good idea?

mslusarczyk commented:
PR checks are failing

mslusarczyk commented:
@kzantow could you confirm that 'skip' button was removed?

kzantow commented:
@mslusarczyk seems it was in error, I have some updates to this PR to address your feedback that will get pushed shortly

mslusarczyk commented on war/src/main/js/pluginSetupWizardGui.js:
```diff
@@ -373,36 +398,37 @@ var createPluginSetupWizard = function(appendTarget) {
+		DEFAULT: function() {
+			setPanel(welcomePanel);
+			// focus on default
+			$('.install-recommended').focus();
```
can welcomePanel name be overwritten? if so this will result in an exception.

kzantow replied:
No, welcomePanel is a local variable. But the DEFAULT state can be overwritten.

mslusarczyk replied:
So what UpgradeWizard.java:55 is setting? Is it still used?

kzantow replied:
No, I'm getting rid of that, good catch. Simplifying things as much as I can.

mslusarczyk commented on core/src/main/java/jenkins/install/InstallStateFilter.java:
```diff
@@ -15,7 +15,7 @@
     /**
```
Class javadoc is invalid there is no InstallUtil#getInstallState().

mslusarczyk commented on war/src/main/js/templates/pluginSelectionPanel.hbs:
```diff
@@ -24,49 +24,17 @@
 		{{translations.installWizard_goBack}}
 	</button>
-	<button type="button" class="btn btn-primary install-selected">
+	<button id="install-plugins-button" type="button" class="btn btn-primary install-selected" {{#unless selectedPlugins}}disabled="disabled"{{/unless}}>
```
🐛 doesn't allow me to install 'none' plugins.

kzantow replied:
You can click the X to close the wizard

mslusarczyk replied:
Right, but I'm skipping 'Create Admin User' this way also, correct?

kzantow replied:
Good point, was mostly hoping to prevent accidentally clicking 'next' there

mslusarczyk commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -1313,20 +1311,18 @@ public HttpResponse doInstallPlugins(StaplerRequest req) throws IOException {
-        updateCenter.persistInstallStatus();
-        if (jenkins.getInstallState() == InstallState.NEW) {
+        if (!Jenkins.getInstance().getInstallState().isSetupComplete()) {
             jenkins.setInstallState(InstallState.INITIAL_PLUGINS_INSTALLING);
```
hardcoded transition to INITIAL_PLUGINS_INSTALLING state.

kzantow replied:
Right, this one should be. At this point, initial plugins are actually installing and providing feedback to the user.

mslusarczyk replied:
OK, so why do we need separate NEW and INITIAL_PLUGINS_INSTALLING states if we are  moving between them outside of InstallState flow.

kzantow replied:
you might have different states between those states. this also allows refreshing the page to display the correct panel, regardless of the state you are currently on.

mslusarczyk replied:
This is fine after INITIAL_SECURITY_SETUP was added.

mslusarczyk commented on core/src/main/java/jenkins/install/InstallUtil.java:
```diff
@@ -70,12 +75,76 @@
+                if (current == null || InstallState.UNKNOWN.equals(current)) {
+                    return getDefaultInstallState();
+                }
+                final Map<InstallState, InstallState> states = new HashMap<InstallState, InstallState>();
```
this should be private static final in class

kzantow replied:
I'm not a fan of static state & a little memory can be reclaimed this way.

mslusarczyk commented on core/src/main/java/jenkins/install/SetupWizard.java:
```diff
@@ -61,89 +66,94 @@
+            try {
+                PluginServletFilter.addFilter(FORCE_SETUP_WIZARD_FILTER);
+            } catch (ServletException e) {
+                throw new AssertionError(e);
```
why AssertionError? Also message detailing the exc would be nice.

mslusarczyk commented on war/pom.xml:
```diff
@@ -676,6 +676,11 @@ THE SOFTWARE.
       <build>
+        <resources>
+          <resource>
+            <directory>${project.build.directory}/generated-resources/frontend</directory>
```
is it still used? nothing is generated there.

mslusarczyk commented:
🐝

mslusarczyk commented on core/src/main/java/jenkins/install/UpgradeWizard.java:
```diff
@@ -41,13 +42,43 @@
+        // If there are no platform updates, proceed to running
+        if (isUpToDate) {
+            if (Jenkins.getInstance().getSetupWizard().getPlatformPluginUpdates().isEmpty()) {
+                Jenkins.getInstance().setInstallState(InstallState.RUNNING);
```
in InstallUtil there is states.put(InstallState.UPGRADE, InstallState.INITIAL_SETUP_COMPLETED);

why RUNNING is forced here?

kzantow replied:
because this is basically a restart with no valid upgrade, which would be "RESTART" or "RUNNING"

mslusarczyk commented on core/src/main/java/jenkins/install/SetupWizard.java:
```diff
@@ -404,6 +405,16 @@ void completeSetup() throws IOException {
     /**
+     * Returns an installState by name
+     */
+    public InstallState getInstallState(String name) {
```
unused method.

mslusarczyk commented on core/src/main/java/jenkins/install/SetupWizard.java:
```diff
@@ -404,6 +405,16 @@ void completeSetup() throws IOException {
+     * Returns an installState by name
+     */
+    public InstallState getInstallState(String name) {
+        if (name == null) {
```
could be a part of InstallState.valueOf

mslusarczyk commented on core/src/main/java/jenkins/install/SetupWizard.java:
```diff
@@ -374,28 +379,39 @@ public HttpResponse doPlatformPluginList() throws IOException {
+    /**
+     * Returns an installState by name
+     */
+    public InstallState getInstallState(String name) {
```
unused method

kzantow replied:
this is used from stapler

mslusarczyk commented on core/src/main/java/jenkins/install/InstallUtil.java:
```diff
@@ -70,12 +75,77 @@
+        if (!current.equals(prior)) {
+            LOGGER.warning("Transitioning state from: " + prior + ", but current is: " + current);
+        }
+        InstallState next = getNextInstallState(prior);
```
does that overwrites 'current'?  why do we need to define 'prior'?

mslusarczyk commented:
not a part of latest changes but PluginManager.java:673 should use equals.

mslusarczyk commented:
🐛 after the wizard is completed and Jenkins restarted I still see Unlock Jenkins screen. Seems that filter redirecting to it is not removed.

daniel-beck commented:
Findings in PR build 5255:

* If after upgrading one installs only the Pipeline plugin (leaving GHOF uninstalled), the banner disappears.
* If the Pipeline plugin was installed before upgrading, it's still offered in the wizard. If the user selects to install it, the progress screen hangs indefinitely.
* The 'Back' button is somewhere completely differently from the 'Continue anyway' button when skipping installation, and basically impossible to notice on a large screen.
* The 'X' button in the wizard behaves differently from the 'X' button in the banner, the latter should be replaced by something different.

mslusarczyk commented:
still a 🐛. Important part to reproducing is to continue as a auto-generated admin on 'Create First Admin User' screen

kzantow commented:
@daniel-beck your comments have been addressed here, please give the latest build a spin! :)
@mslusarczyk I've corrected the issue when 'continuing as admin' please let me know if there's anything else outstanding for your uses

recena commented on war/src/main/webapp/css/style.css:
```diff
@@ -1859,6 +1859,18 @@ table#legend-table td {
+body.no-decoration #side-panel,
+body.no-decoration footer {
+    display: none;
+}
```
All of this will be removed when [JENKINS-34670](https://issues.jenkins-ci.org/browse/JENKINS-34670) is implemented.

kzantow replied:
right, that's the idea @recena

daniel-beck commented:
Build 5263: For some reason, the plugins in the wizard only show their ID for me, no description, and when I go to install them, it says

>  An error occurred during installation: No such plugin: workflow-aggregator 

FWIW plugins not found in the update site must never be suggested -- this will be relevant once plugins start depending on Jenkins 2.x (x ≠ 0), but users still install 2.0. Inapplicable suggestions (including those plugins marked as incompatible as either they or a dependency requires a newer core) must not be presented in upgrade or install wizards.

Otherwise, the 'minimize' button is a great improvement, and the 'no thanks' wording is much nicer.

daniel-beck commented:
* Doesn't offer to install things that are already installed.
* Doesn't offer to install things if upgrading from 2.x and not having them installed.
* Doesn't show when everything that would be offered is already installed.

Very nice!

Making sure not to offer something not on the update site, or not compatible with the installed version, would be the only thing I noticed that would be left to do. (This can also happen with LTS releases, as everyone using 1.642.2 knows -- Pipeline currently requires 1.642.3 and shows as incompatible on .2)

mslusarczyk commented:
🐝

rtyler commented on core/src/main/resources/jenkins/install/pluginSetupWizard.properties:
```diff
@@ -56,3 +58,20 @@ installWizard_pluginInstallFailure_message=Some plugins failed to install proper
+installWizard_upgradeSkipped_banner=Features Not Installed
+installWizard_upgradeSkipped_message=<div class="alert alert-warning fade in">No new plugins will be installed</div> \
+<p style="padding: 0 4px">You can also install new features from the Plugin Manager, if you change your mind.</p> \
+<p style="padding: 0 4px">Visit the <a target="_blank" href="https://jenkins.io/2.0/">Jenkins 2 website</a> \
```
> Visit the Jenkins 2 website to learn how these new features can improve your Jenkins experience.

This sentence strikes me as awkward but I'm having difficulty explaining what it is exactly. If this were in documentation or a page on the site, I generally try to end the "call to action" with the link. I think that kind of sentence structure is more likely to get somebody to click, since they're not visually back-tracking to click the right link (perhaps I spend too much time thinking about these kinds of things :smile: )

I suggest:

> Learn how these new features can improve your Jenkins experience on the **Jenkins 2 homepage**

rtyler commented on core/src/main/resources/jenkins/install/pluginSetupWizard.properties:
```diff
@@ -56,3 +58,20 @@ installWizard_pluginInstallFailure_message=Some plugins failed to install proper
+installWizard_upgradeComplete_message=Congratulations! You have upgraded to Jenkins {0}.</p><p>To learn more about the great new features, visit <a target="_blank" href="https://jenkins.io/2.0/">the Jenkins website</a>!
+installWizard_upgradeSkipped_title=Upgrade
+installWizard_upgradeSkipped_banner=Features Not Installed
+installWizard_upgradeSkipped_message=<div class="alert alert-warning fade in">No new plugins will be installed</div> \
```
> No new plugins will be installed

I think that this could be confusing, I wouldn't want somebody to assume that the ability to install new plugins has been disabled.

I suggest:

> Suggested plugins will not be installed.

rtyler commented on core/src/main/resources/jenkins/install/pluginSetupWizard.properties:
```diff
@@ -56,3 +58,20 @@ installWizard_pluginInstallFailure_message=Some plugins failed to install proper
+installWizard_upgradePanel_skipRecommendedPlugins=No thanks
+installWizard_upgradeComplete_title=Upgrade
+installWizard_pluginsInstalled_banner=Welcome to Jenkins {0}!
+installWizard_upgradeComplete_message=Congratulations! You have upgraded to Jenkins {0}.</p><p>To learn more about the great new features, visit <a target="_blank" href="https://jenkins.io/2.0/">the Jenkins website</a>!
```
> Congratulations! You have upgraded to Jenkins {0}.To learn more about the great new features

IMO the last sentence should be more possessive

> Congratulations! You have upgraded to Jenkins {0}. To learn more about ***its***  great new features, visit foo

rtyler commented on core/src/main/resources/jenkins/install/pluginSetupWizard.properties:
```diff
@@ -56,3 +58,20 @@ installWizard_pluginInstallFailure_message=Some plugins failed to install proper
+installWizard_upgradePanel_banner=Welcome to Jenkins {0}!
+installWizard_upgradePanel_message=Jenkins {0} includes some great new features that we think you'll love, install these additional plugins to take advantage of them!
+installWizard_upgradePanel_recommendedActionTitle=Install Suggested Plugins
+installWizard_upgradePanel_recommendedActionDetails=Install the suggested plugins
```
The redundancy here is somewhat goofy but I don't have a better suggestion at this time.

kzantow commented:
@daniel-beck I've added a test for a compatible version of the suggested plugins here https://github.com/jenkinsci/jenkins/pull/2281/commits/ad454d3472159c993273ceabdd5ecfa6ae722386#diff-f65b8a70854ca1cc6c12397eee54d279R370

mslusarczyk commented:
LGTM 🐝

daniel-beck commented:
Seems :+1:

oleg-nenashev commented:
After manual tests the behavior looks good to me. Hence 🐝 from me (I'm not the best code reviewer for this stuff BTW).

recena commented:
Me too, :bee: some details can be polish later. Great work!

kzantow commented:
*sigh* ... Finished: ABORTED

kzantow committed:
[FIX JENKINS-33663] - Upgrade wizard

Add Upgrade Wizard
Add 'replace' handlebars method in jenkins.js
Fix state transitions to follow a single basic pattern
Suggested plugins selected by default
Add an install state for INITIAL_SECURITY_SETUP
Removed duplicate setupWizard.js on page [FIX JENKINS-34676]
Exclude plugins which are already installed when determining platform
Change 'snooze' to not look like a 'close' button
Make sure unlock screen doesn't show up after new install
Add compatibility check when offering suggested plugins
More appropriate handling of security token filter
